### PR TITLE
Bed 3922 - Zip File Upload Support

### PR DESCRIPTION
**Pull request by rvazarkar**
## Description

This MR enables consumption of ZIP files by the BloodHound API. 

## Motivation and Context
Zip file upload is a feature that has been repeatedly requested and serves the use case of BloodHound quite well. This re-enables a feature that Legacy already had.

## How Has This Been Tested?
Integration tests have been expanded to cover zip file ingestion

## Types of changes

-   [ ] Chore (a change that does not modify the application functionality)
-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

-   [x] Documentation updates are needed, and have been made accordingly.
-   [x] I have added and/or updated tests to cover my changes.
-   [x] All new and existing tests passed.
-   [x] My changes include a database migration.

### CODE REVIEW

File: cmd/api/src/api/v2/apiclient/apiclient.go [hunk 1]
```diff
@@ -78,6 +78,51 @@ func ReadAPIError(response *http.Response) error {
...
+				request.Body = io.NopCloser(bytes.NewReader(body))
+			}
+
+			// Set our credentials either via signage or bearer token session
+			// Credentials also have to be set with every attempt due to request signing
+			if s.Credentials != nil {
+				if err := s.Credentials.Handle(request); err != nil {
+					return nil, err
+				}
+			}
```
Comment (definitelynotagoblin) [inline cmd/api/src/api/v2/apiclient/apiclient.go:110]: Is there any danger of leaving the request body reader open if we fail out of ZipRequest due to bad credentials?

File: cmd/api/src/api/v2/apiclient/ingest.go [hunk 2]
```diff
@@ -89,7 +104,7 @@ func (s Client) SendCompressedFileUploadData(jsonFile io.Reader, id int64) error
 		return fmt.Errorf("failed to close gzip writer: %w", err)
 	}
 
-	request, err := s.NewRequest(http.MethodPost, fmt.Sprintf("api/v2/file-upload/%d", id), nil, io.NopCloser(&body))
+	request, err := s.NewRequest(http.MethodPost, fmt.Sprintf("api/v2/file-upload/%d", id), nil, io.NopCloser(&body), http.Header{headers.ContentType.String(): []string{mediatypes.ApplicationJson.String()}})
```
Comment (mistahj67) [inline cmd/api/src/api/v2/apiclient/ingest.go:107]: shouldn't this be the mediatype for gzip?

File: cmd/api/src/api/v2/file_uploads.go [hunk 3]
```diff
@@ -130,15 +131,17 @@ func (s Resources) ProcessFileUpload(response http.ResponseWriter, request *http
 		fileUploadJobIdString = mux.Vars(request)[FileUploadJobIdPathParameterName]
 	)
 
-	if fileUploadJobID, err := strconv.Atoi(fileUploadJobIdString); err != nil {
+	if !api.HeaderMatches(headers.ContentType.String(), mediatypes.ApplicationJson.String(), request.Header) && !api.HeaderMatches(headers.ContentType.String(), mediatypes.ApplicationZip.String(), request.Header) {
```
Comment (mistahj67) [inline cmd/api/src/api/v2/file_uploads.go:134]: There's a constant for accepted types `bhce/cmd/api/src/api/v2/file_uploads.go`. I think it would be a good idea to rely on that constant here if possible.

File: cmd/api/src/daemons/datapipe/jobs.go [hunk 4]
```diff
@@ -119,26 +121,69 @@ func (s *Daemon) clearFileTask(ingestTask model.IngestTask) {
...
 	}
 }
 
+func (s *Daemon) processIngestFile(ctx context.Context, path string, fileType model.FileType) error {
+	paths, err := s.preProcessIngestFile(path, fileType)
+
+	if err != nil {
+		return err
+	}
+
```
Comment (mistahj67) [inline cmd/api/src/daemons/datapipe/jobs.go:169]: I think this falls under the convention of moving the err check up to an if check right?

File: cmd/api/src/go.mod [hunk 5]
```diff
@@ -47,7 +47,9 @@ require (
 	github.com/unrolled/secure v1.13.0
 	github.com/zenazn/goji v1.0.1
 	go.uber.org/mock v0.2.0
+	go.uber.org/zap v1.13.0
 	golang.org/x/crypto v0.18.0
+	golang.org/x/exp v0.0.0-20230515195305-f3d0a9c9a5cc
```
Comment (mistahj67) [inline cmd/api/src/go.mod:52]: are these intentionally added? I've seen these in other PRs as well so curious if this is need or should be in a different PR?

File: cmd/api/src/daemons/datapipe/jobs.go [hunk 6]
```diff
@@ -119,26 +121,69 @@ func (s *Daemon) clearFileTask(ingestTask model.IngestTask) {
...
+				filePaths[i] = tempFile.Name()
 			}
-		})
+		}
+
+		//Close the archive and delete it
+		if err := archive.Close(); err != nil {
+			log.Errorf("Error closing archive %s: %v", path, err)
+		} else if err := os.Remove(path); err != nil {
+			log.Errorf("Error deleting archive %s: %v", path, err)
```
Comment (definitelynotagoblin) [inline cmd/api/src/daemons/datapipe/jobs.go:156]: This is probably correct behavior but just confirming that we want to delete archives even if they partially fail to process.

File: cmd/api/src/model/ingest/ingest.go [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+	CollectionMethodSessionLoop   CollectionMethod = 1 << 10
+	CollectionMethodLoggedOnLoop  CollectionMethod = 1 << 11
+	CollectionMethodDCOM          CollectionMethod = 1 << 12
+	CollectionMethodSPNTargets    CollectionMethod = 1 << 13
+	CollectionMethodPSRemote      CollectionMethod = 1 << 14
+	CollectionMethodUserRights    CollectionMethod = 1 << 15
+	CollectionMethodCARegistry    CollectionMethod = 1 << 16
+	CollectionMethodDCRegistry    CollectionMethod = 1 << 17
+	CollectionMethodCertServices  CollectionMethod = 1 << 18
+)
```
Comment (mistahj67) [inline cmd/api/src/model/ingest/ingest.go:132]:
```suggestion
const (
	CollectionMethodGroup CollectionMethod = 1 << iota
	CollectionMethodLocalAdmin
	CollectionMethodGPOLocalGroup
	CollectionMethodSession
	CollectionMethodLoggedOn
	CollectionMethodTrusts
	CollectionMethodACL
	CollectionMethodContainer
	CollectionMethodRDP
	CollectionMethodObjectProps
	CollectionMethodSessionLoop
	CollectionMethodLoggedOnLoop
	CollectionMethodDCOM
	CollectionMethodSPNTargets
	CollectionMethodPSRemote
	CollectionMethodUserRights
	CollectionMethodCARegistry
	CollectionMethodDCRegistry
	CollectionMethodCertServices
)
```

File: cmd/api/src/model/ingest.go [hunk 8]
```diff
@@ -1,43 +1,39 @@
...
-	if err := os.Remove(s.FileName); err != nil {
-		log.Errorf("Error removing ingest file %v: %v", s.FileName, err)
-	}
+type FileType int
 
-	return nil
-}
+const (
+	FileTypeJson FileType = 0
+	FileTypeZip  FileType = 1
```
Comment (mistahj67) [inline cmd/api/src/model/ingest.go:38]:
```suggestion
	FileTypeJson FileType = iota
	FileTypeZip
```

**Review by mistahj67 — COMMENTED** [review]
Did we want to incorporate the copyWithCtx for the io.Copy's into this PR from @computator ?

File: cmd/api/src/api/v2/apiclient/ingest.go (hunk 2 above)
Comment (rvazarkar) [reply cmd/api/src/api/v2/apiclient/ingest.go:107]: Nope, zip file != gzip file

File: cmd/api/src/api/v2/apiclient/apiclient.go (hunk 1 above)
Comment (rvazarkar) [reply cmd/api/src/api/v2/apiclient/apiclient.go:110]: I'm not 100% positive, but this is all in integration tests so not the biggest deal

File: cmd/api/src/services/fileupload/validation.go [hunk 9]
```diff
@@ -14,49 +14,50 @@
...
 		metaTagFound     = false
-		meta             Metadata
+		meta             ingest.Metadata
 	)
 
 	for {
 		if dataTagValidated && metaTagFound {
-			return meta, nil
+			break
 		}
```
Comment (mistahj67) [inline cmd/api/src/services/fileupload/validation.go:44]: This pattern took me a second to parse that the success kicks out in the subsequent loop iteration. What do you think of moving this to the end of the for loop for readability?

File: cmd/api/src/services/ingest/ingest.go [hunk 10]
```diff
@@ -25,11 +25,12 @@ type IngestData interface {
 	CreateIngestTask(task model.IngestTask) (model.IngestTask, error)
 }
 
-func CreateIngestTask(db IngestData, filename string, requestID string, jobID int64) (model.IngestTask, error) {
+func CreateIngestTask(db IngestData, filename string, requestID string, jobID int64, fileType model.FileType) (model.IngestTask, error) {
```
Comment (mistahj67) [inline cmd/api/src/services/ingest/ingest.go:28]: Do we have any conventions on argument order? Seems like fileName would go with fileType unless we are grouping by variable type 🤔 in which case we can remove the first `string` declaration right?

**Review by definitelynotagoblin — APPROVED** [review]

File: cmd/api/src/api/v2/apiclient/apiclient.go [hunk 11]
```diff
@@ -134,11 +179,17 @@ func (s Client) Request(method, path string, params url.Values, body any, header
...
+func (s Client) NewRequest(method string, path string, params url.Values, body io.ReadCloser, header ...http.Header) (*http.Request, error) {
 	endpoint := api.URLJoinPath(s.ServiceURL, path)
 	endpoint.RawQuery = params.Encode()
 
-	return http.NewRequest(method, endpoint.String(), body)
+	request, err := http.NewRequest(method, endpoint.String(), body)
+
+	if len(header) > 0 {
+		request.Header = header[0]
+	}
```
[the remark this answers is not in the document]
Comment (rvazarkar) [reply cmd/api/src/api/v2/apiclient/apiclient.go:190]: Its just integration tests, so I'm not super concerned about it for now, I copied this from an existing example

File: cmd/api/src/services/fileupload/file_upload.go [hunk 12]
```diff
@@ -147,6 +131,24 @@ func SaveIngestFile(location string, request *http.Request) (string, model.FileT
 	}
 }
 
+type FileValidator func(src io.Reader, dst io.Writer) error
```
Comment (mistahj67) [inline cmd/api/src/services/fileupload/file_upload.go:134]: If the WriteAndValidateFile function returned the filename, modeType, and error you'd be able to eliminate the if / else checks above as well correct?

File: cmd/api/src/go.mod [hunk 13]
```diff
@@ -48,6 +48,7 @@ require (
 	github.com/zenazn/goji v1.0.1
 	go.uber.org/mock v0.2.0
 	golang.org/x/crypto v0.18.0
+	golang.org/x/exp v0.0.0-20230515195305-f3d0a9c9a5cc
```
Comment (mistahj67) [inline cmd/api/src/go.mod:51]: Did we intentionally want this dependency added?

**Review by mistahj67 — APPROVED** [review]
I can't comment to much on the validation logic but from what I can comment on, LGTM sir ✅ 

Just verify we want that dependency added before 🚀 please 🙇

File: cmd/api/src/go.mod (hunk 13 above)
Comment (rvazarkar) [reply cmd/api/src/go.mod]: Removed
